### ISSUE TITLE
Some linux distros fail to run appimage due to sandboxing

### ISSUE THREAD

**Issue by lane-c-wagner** [Q-Vault/qvault#346]
Reported in the discord:
```
write(2, "[17211:0822/153126.805948:FATAL:"..., 
292[17211:0822/153126.805948:FATAL:setuid_sandbox_host.cc(157)] The SUID sandbox helper binary
 was found, but is not configured correctly. Rather than run without sandboxing I'm aborting now. 
You need to make sure that /tmp/.mount_QVaultD8sOM8/chrome-sandbox is owned by root and has
 mode 4755.
```
There are known issues about this in electron and electron-builder. We may want to wait until they solve them.

**Comment by preston-wagner** [comment]
I've tested and can't reproduce this on ubuntu or lubuntu; looking at the electron issue, it would seem that the issue is exclusive to linux distros that don't support unprivileged CLONE_NEWUSER (of which arch linux is one); it seems that the only workaround is disabling the electron sandbox, which I feel like is an unneccesary security risk.

**Comment by eli-schwartz** [comment]
Hi, official @archlinux representative here.

I've just been helping one of our users create a distribution package for your software: https://aur.archlinux.org/packages/qvault/

And I noticed this issue in the process. So I would like to point out that the CLONE_NEWUSER stuff is a bit of a red herring: while it is true that electron requires EITHER the `chrome-sandbox` binary to be setuid root, OR for the kernel to permit unprivileged CLONE_NEWUSER (or for both to be true), the real answer to the problem "well, what do people who write electron apps do, then" is, electron apps are expected to use the system electron.

We provide our own versions of electron already: https://www.archlinux.org/packages/?q=electron
We have electron 6 available, as well as compat packages for electron2 and electron4. These packages depend properly on their libxss, gtk3, nss, gconf, etc. dependencies (the prebuilt electron assumes that some packages are installed on a Linux desktop, which "mostly works" because "most people use gnome" and therefore have gtk3, and "probably have firefox or chromium installed", and therefore have nss installed), and for electron 6 and on, we do and will continue to provide a correctly setuid `chrome-sandbox`. The electron package is additionally built from source, uses system libraries where appropriate, and doesn't require every single electron application the user has installed, to maintain numerous private copies of electron.

Using the system electron is very simple: `electron /path/to/app.asar`. Here is the change that implemented the Arch Linux qvault package as a properly conforming system-electron package: https://aur.archlinux.org/cgit/aur.git/commit/?h=qvault&id=b82f7f1c741a7042da1b8504a37a9edfabf81b5e

If you could make this process easier, that would probably be helpful. A build target that creates the resources/app.asar without relying on electron-builder could come in handy.

**Comment by lane-c-wagner** [comment]
Thanks for reaching out! 

So to be clear you would like if there was a way to generate the .asar file without having to depend on electron-builder? The other way I'm seeing to create the asar file more directly would be here: https://electronjs.org/docs/tutorial/application-packaging

However, this introduces another dependency on the asar npm package. Does that even help at all?

**Comment by eli-schwartz** [comment]
The main source of confusion is that you need to copy over the `files` (main.js, index.html, dist/, main/) plus package.json, but you also need to copy over some of the node_modules. Determining exactly which ones you need is non-trivial, and results in needing to use electron-builder at the moment.

Whether you pack it into an asar or not, probably does not matter much as electron can run it either way.

It would be nice if electron-builder had some upstream tooling to divorce the resources/app/ generation from the electron parts, with the ultimate goal of not needing to download so many large binary packages from the NPM registry during a scratch build. This is the cause of much of the complexity of the AUR package now -- it patches package.json to remove electron from `dependencies`, patches in the `electronDist` key to electron-builder's configuration to trick it into cp'ing the system copy in /usr/lib/electron, and then passes some magic arguments to `yarn release` so that appimages and snaps aren't created (because all that is needed is linux-unpacked, and only a small portion of that).

My hope is that eventually the state of electron in linux distributions will become widespread, and it will no longer be recommended in the electron community to vendor very large browser components into everyone's applications, but, we'll see how that goes...

It seems like something that electron-builder and indeed the electron developers themselves should put a high priority on (integrating with linux distributions), because the electron>=6 chrome-sandbox is kind of a bad situation. :D

**Comment by eli-schwartz** [comment]
tl;dr Electron the ecosystem really needs better ways of separating the build steps:

- create an app
- package the app into a prebuilt copy of electron

**Comment by lane-c-wagner** [comment]
I agree that it would be nice to have a stronger decoupling there, but at the same time that means if an app wants electron v6.0.1 and the user has electron v6.1.1 it could lead to some hairy bugs.

I'm also concerned about electron being removed as an npm dependency in the PKGBUILD because there is a lot of runtime code in the app that uses the electron package to do more than just run the app's browser. For example, we use the electron package to communicate between renderer and main processes at runtime. Are you guys sure that using the system electron didn't break that runtime functionality?
 (This is probably our fault for being misleading, electron should be in "dependencies" instead of "devDependencies") 
However, if electron-builder DOES correctly link the runtime electron dependencies then we can help you by including a yarn script that builds a release using the local electron, and omits electron from the download. Let me know on that.

Also, I think I see the magic line you are referring to: 
```
yarn release -l dir -c.electronVersion
```
If you want to provide me with the magical `electronVersion` config, we can add a yarn script for this line so you can just call that script.

Thanks again!

**Comment by XdaemonX** [comment]
Electron has been added as runtime dependency inside PKGBUILD, also tested it right now and it works ok.

**Comment by lane-c-wagner** [comment]
Would this help? https://github.com/Q-Vault/qvault/pull/370

I added a yarn script that handles the entire arch release (as you have it currently). It: 

* patches the package.json file
* installs dependencies
* builds the release
* restores code to original status

The WIP right now is using `--publish never` but if you supply me with the config (c.electronVersion) you are using then I can incorporate that as well.

**Comment by XdaemonX** [comment]
Have tested it right now, it is working ok. However, instead of ```--publish never``` I used ```-l dir -c.electronVersion```. We fake builder so we really don't need to specify version.

**Comment by lane-c-wagner** [comment]
Updated the PR. Do one of you have permissions to mark it approved?
